### PR TITLE
Fix sliders being incorrectly tracked by the wrong keypress

### PR DESCRIPTION
**Pull request by nyquillerium**
Closes https://github.com/ppy/osu/issues/2647

Re-written to perform like stable and now has tests for slider input behavior.

### CODE REVIEW

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+namespace osu.Game.Rulesets.Osu.Tests
+{
+    public class TestCaseSliderInput : SliderTestBase
+    {
+        private readonly Container content;
+        protected override Container<Drawable> Content => content;
+
+        protected override List<Mod> Mods { get; set; }
+
+        private int testID;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: why is this required?

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        ///     At 250ms intervals:
+        ///     Frame 1 (prior to slider):          Left Click
+        ///     Frame 2 (within slider hit window): Left & Right Click
+        ///     Frame 3 (while tracking):           Left Click
+        ///     A passing test case will have the cursor lose tracking on frame 3.
+        /// </summary>
+        [Test]
+        public void TestLeftBeforeSliderThenRight()
+        {
+            AddStep("Invalid key transfer test", () => performInvalidKeyTransferTest());
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: move actual test logic into this method.

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        ///     At 250ms intervals:
+        ///     Frame 1 (prior to slider):          Left Click
+        ///     Frame 2 (within slider hit window): Left & Right Click
+        ///     Frame 3 (while tracking):           Right Click
+        ///     A passing test case will have the cursor continue to track after frame 3.
+        /// </summary>
+        [Test]
+        public void TestLeftBeforeSliderThenRightThenLettingGoOfLeft()
+        {
+            AddStep("Left to both to right test", () => performLeftToRightTransferTest());
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: move actual test logic into this method.

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        ///     At 250ms intervals:
+        ///     Frame 1: Left Click
+        ///     Frame 2: Left & Right Click
+        ///     Frame 3: Right Click
+        ///     A passing test case will have the cursor continue to track after frame 3.
+        /// </summary>
+        [Test]
+        public void TestTrackingRetentionLeftRightLeft()
+        {
+            AddStep("Tracking retention test", () => performTrackingRetentionTest());
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: move actual test logic into this method.

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    {
+        private readonly Container content;
+        protected override Container<Drawable> Content => content;
+
+        protected override List<Mod> Mods { get; set; }
+
+        private int testID;
+        private Player player;
+        private JudgementResult lastResult;
+        private bool allJudgedFired;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: this likely isn't being reset after each test.

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        }
+
+        private bool assertGreatJudge()
+        {
+            return lastResult.Type == HitResult.Great;
+        }
+
+        private void performInvalidKeyTransferTest()
+        {
+            var actions = new List<List<OsuAction>>();
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]:
```csharp 
            var actions = new List<List<OsuAction>>
            {
                new List<OsuAction> { OsuAction.LeftButton },
                new List<OsuAction> { OsuAction.LeftButton, OsuAction.RightButton },
                new List<OsuAction> { OsuAction.LeftButton }
            };
```

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+                    {
+                        Position = sliderToAdd.Position,
+                        Time = sliderToAdd.StartTime + (testInterval * (primeKey ? localIndex - 1: localIndex)),
+                        Actions = actions[localIndex++]
+                    });
+                }
+            }
+
+            thisReplay.Frames = frames;
+            var newPlayer = loadPlayerFor(ruleset, sliderToAdd, thisReplay, beatmap);
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: `newPlayer` pointless variable.

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs (hunk 5 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: okay, it is, but it's done in a place that is very non-obvious. i'd prefer it was done in test's `[SetUp]`.

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 8]
```diff
@@ -147,23 +151,59 @@ private set
 
         private bool canCurrentlyTrack => Time.Current >= slider.StartTime && Time.Current < slider.EndTime;
 
+        private OsuAction? trackingAction;
+
+        private bool cursorTrackingBall()
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: property, not method

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 9]
```diff
@@ -147,23 +151,59 @@ private set
...
+            switch (action)
+            {
+                case OsuAction.LeftButton:
+                case OsuAction.RightButton:
+                    hit = cursorTrackingBall() && (Hit?.Invoke() ?? false);
+                break;
+            }
+
+            if (hit)
+            {
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: unnecessary brackets

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 10]
```diff
@@ -147,23 +151,59 @@ private set
...
+        public bool OnPressed(OsuAction action)
+        {
+            bool hit = false;
+
+            switch (action)
+            {
+                case OsuAction.LeftButton:
+                case OsuAction.RightButton:
+                    hit = cursorTrackingBall() && (Hit?.Invoke() ?? false);
+                break;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: wrong indentation.

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 11]
```diff
@@ -147,23 +151,59 @@ private set
...
+
+        public bool OnPressed(OsuAction action)
+        {
+            bool hit = false;
+
+            switch (action)
+            {
+                case OsuAction.LeftButton:
+                case OsuAction.RightButton:
+                    hit = cursorTrackingBall() && (Hit?.Invoke() ?? false);
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: don't need `bool hit`

```csharp
if (...)
{
	trackingAction = action;
	return true;
}
```

File: osu.Game.Rulesets.Osu/Replays/OsuAutoGenerator.cs [hunk 12]
```diff
@@ -106,7 +106,7 @@ private void addDelayedMovements(OsuHitObject h, OsuHitObject prev)
             }
         }
 
-        private void addHitObjectReplay(OsuHitObject h)
+        protected void addHitObjectReplay(OsuHitObject h)
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Replays/OsuAutoGenerator.cs]: why?

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+            player = newPlayer;
+            Child = player;
+
+            ((ScoreAccessibleReplayPlayer)player).ScoreProcessor.NewJudgement += result =>
+            {
+                lastResult = result;
+                Logger.Log(result.Type.ToString());
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: remove unnecessary logging (use the test steps to provide output on tests)

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+            var sliderToAdd = CreateSlider(distance: 5, addToContent: false);
+            Ruleset ruleset = new OsuRuleset();
+            var beatmap = createBeatmap(sliderToAdd, ruleset);
+
+            var thisReplay = new Replay();
+            var frames = new List<ReplayFrame>();
+            const double testInterval = 250;
+            var localIndex = 0;
+
+            //Empty frame to be added as a workaround for first frame behavior.
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: space after `//`

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+                        Actions = actions[localIndex++]
+                    });
+                }
+                else
+                {
+                    frames.Add(new OsuReplayFrame
+                    {
+                        Position = sliderToAdd.Position,
+                        Time = sliderToAdd.StartTime + (testInterval * (primeKey ? localIndex - 1: localIndex)),
+                        Actions = actions[localIndex++]
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: just use `a`?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Replays/OsuAutoGenerator.cs (hunk 12 above)
Comment (nyquillerium) [reply osu.Game.Rulesets.Osu/Replays/OsuAutoGenerator.cs]: was accidentally left in after testing

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs (hunk 1 above)
Comment (nyquillerium) [reply osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: was accidentally left in for testing, will remove

File: osu.Game.Rulesets.Osu.Tests/SliderTestBase.cs [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Objects.Types;
+using osu.Game.Rulesets.Osu.Objects.Drawables.Pieces;
+
+namespace osu.Game.Rulesets.Osu.Tests
+{
+    [TestFixture]
+    public abstract class SliderTestBase : OsuTestCase
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/SliderTestBase.cs]: Revert this change (move back to TestCaseSlider). You don't need to inherit from this (just make the slider locally).

File: osu.Game.Rulesets.Osu.Tests/TestCaseSlider.cs [hunk 17]
```diff
@@ -1,53 +1,27 @@
...
-using osu.Game.Rulesets.Osu.Objects.Drawables;
-using osu.Game.Tests.Visual;
 using osuTK;
-using osuTK.Graphics;
 using osu.Game.Rulesets.Mods;
-using System.Linq;
 using NUnit.Framework;
-using osu.Game.Graphics.Sprites;
-using osu.Game.Rulesets.Judgements;
+using osu.Framework.Graphics;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSlider.cs]: Revert all changes in this file.

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+using osu.Game.Rulesets.Osu.Replays;
+using osu.Game.Rulesets.Replays;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Scoring;
+using osu.Game.Screens.Play;
+using osu.Game.Tests.Beatmaps;
+
+namespace osu.Game.Rulesets.Osu.Tests
+{
+    public class TestCaseSliderInput : SliderTestBase
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: `: OsuTestCase`

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+        ///     Frame 2 (within slider hit window): Left & Right Click
+        ///     Frame 3 (while tracking):           Left Click
+        ///     A passing test case will have the cursor lose tracking on frame 3.
+        /// </summary>
+        [Test]
+        public void TestLeftBeforeSliderThenRight()
+        {
+            AddStep("Invalid key transfer test", () =>
+            {
+                var actions = new List<List<OsuAction>>();
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: please check my previous review, all these have not been converted to object initialisers.

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+        ///     Hitting a slider head, pressing a new key after the initial hit, then letting go of the original key used to hit
+        ///     the slider should reslt in continued tracking.
+        ///     At 250ms intervals:
+        ///     Frame 1: Left Click
+        ///     Frame 2: Left & Right Click
+        ///     Frame 3: Right Click
+        ///     A passing test case will have the cursor continue to track after frame 3.
+        /// </summary>
+        [SetUp]
+        public void Setup()
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: put at the top of the class

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+                var frame3 = new List<OsuAction> { OsuAction.LeftButton };
+                actions.Add(frame3);
+
+                performStaticInputTest(actions, true);
+            });
+            AddUntilStep(() => allJudgedFired, "Wait for test 1");
+            AddAssert("Tracking lost", assertMehJudge);
+        }
+
+        /// <summary>
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: xmldoc on wrong method

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs (hunk 21 above)
Comment (nyquillerium) [reply osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: uh it looks correct though

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+                Time = 250,
+                Actions = primeKey ? new List<OsuAction> { OsuAction.LeftButton } : new List<OsuAction>()
+            });
+
+            foreach (var a in actionsOnSlider)
+            {
+                frames.Add(new OsuReplayFrame
+                {
+                    Position = sliderToAdd.Position,
+                    Time = sliderToAdd.StartTime + test_interval * actionsOnSlider.IndexOf(a),
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: don't use indexof. keep a time variable

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+        }
+
+        private bool assertGreatJudge()
+        {
+            return lastResult.Type == HitResult.Great;
+        }
+
+        private void performStaticInputTest(List<List<OsuAction>> actionsOnSlider, bool primeKey = false)
+        {
+            var sliderToAdd = new Slider
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: just call this `slider`

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+        [SetUp]
+        public void Setup()
+        {
+            Schedule(() => { allJudgedFired = false; });
+        }
+
+        private readonly Container content;
+        protected override Container<Drawable> Content => content;
+
+        private ScoreAccessibleReplayPlayer player;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: doesn't need to be field.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 25]
```diff
@@ -0,0 +1,240 @@
...
+                frames.Add(new OsuReplayFrame
+                {
+                    Position = slider.Position,
+                    Time = slider.StartTime + frameTime,
+                    Actions = a
+                });
+                frameTime += 250;
+            }
+
+            Beatmap.Value = new TestWorkingBeatmap(new Beatmap<OsuHitObject>
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]:
```csharp
            Beatmap.Value = new TestWorkingBeatmap(new Beatmap<OsuHitObject>
            {
                HitObjects = { slider },
                ControlPointInfo =
                {
                    DifficultyPoints = { new DifficultyControlPoint { SpeedMultiplier = 0.1f } }
                },
                BeatmapInfo =
                {
                    BaseDifficulty = { SliderTickRate = 3 },
                    Ruleset = new OsuRuleset().RulesetInfo
                },
            });
```

feels better

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 26]
```diff
@@ -0,0 +1,240 @@
...
+            {
+                HitObjects = { slider },
+                ControlPointInfo = new ControlPointInfo { DifficultyPoints = { new DifficultyControlPoint { SpeedMultiplier = 0.1f } } },
+                BeatmapInfo = new BeatmapInfo { BaseDifficulty = new BeatmapDifficulty { SliderTickRate = 3 }, Ruleset = new OsuRuleset().RulesetInfo },
+            });
+
+            ScoreAccessibleReplayPlayer player = new ScoreAccessibleReplayPlayer(new Score { Replay = new Replay { Frames = frames } })
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]:
```suggestion
            var player = new ScoreAccessibleReplayPlayer(new Score { Replay = new Replay { Frames = frames } })
```

this is a good example of where `var` should be used.

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs [hunk 27]
```diff
@@ -43,25 +43,34 @@ public DrawableSlider(Slider s)
...
-                    Alpha = 0
-                },
-                HeadCircle = new DrawableSliderHead(s, s.HeadCircle)
-                {
-                    OnShake = Shake
+                    Alpha = 0,
+                    Hit = () =>
+                    {
+                        if (AllJudged)
+                        return false;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs]: wrong indentation

File: osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs [hunk 28]
```diff
@@ -43,25 +43,34 @@ public DrawableSlider(Slider s)
...
+
+                        UpdateResult(true);
+                        return true;
+                    },
                 },
+                HeadCircle.CreateProxy(),
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/DrawableSlider.cs]:
```suggestion
                HeadCircle.CreateProxy(), // required to allow the sliderball to always receive input for judgement purposes
```

should always inline comment when there's a special case. so people know why the special case exists without having to track down git history.

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 29]
```diff
@@ -13,12 +15,14 @@
...
 namespace osu.Game.Rulesets.Osu.Objects.Drawables.Pieces
 {
-    public class SliderBall : CircularContainer, ISliderProgress
+    public class SliderBall : CircularContainer, ISliderProgress, IKeyBindingHandler<OsuAction>
     {
         private const float width = 128;
 
         private Color4 accentColour = Color4.Black;
 
+        public Func<bool> Hit;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: xmldoc. what is the bool return here? when is this fired?

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Behaviour still doesn't match stable.

L down -> R down -> R up while tracking slider causes untracking on lazer, but should not, as one example.

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 30]
```diff
@@ -0,0 +1,240 @@
...
+        ///     Pressing a key before a slider, pressing the other key on the slider head, then releasing the former pressed key
+        ///     should result in continued tracking.
+        ///     At 250ms intervals:
+        ///     Frame 1: Left Click
+        ///     Frame 2: Left & Right Click
+        ///     Frame 3: Right Click
+        ///     A passing test case will have the cursor continue to track after frame 3.
+        /// </summary>
+        [Test]
+        public void TestTrackingLeftBeforeSliderToRight()
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: this test seems to do exactly the same as the one below it?

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 31]
```diff
@@ -147,23 +151,48 @@ private set
...
 
             if (Time.Current < slider.EndTime)
             {
                 // Make sure to use the base version of ReceivePositionalInputAt so that we correctly check the position.
-                Tracking = canCurrentlyTrack
-                           && lastScreenSpaceMousePosition.HasValue
-                           && ReceivePositionalInputAt(lastScreenSpaceMousePosition.Value)
-                           && (drawableSlider?.OsuActionInputManager?.PressedActions.Any(x => x == OsuAction.LeftButton || x == OsuAction.RightButton) ?? false);
+                Tracking = cursorTrackingBall
+                        && canCurrentlyTrack
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: Would reorder this to go `canCurrentlyTrack && cursorTrackingBall`, to save a `ToLocalSpace()` computation.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs (hunk 30 above)
Comment (nyquillerium) [reply osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: As discussed, will make this method more readable by pairing timestamps with inputs

**Comment by nyquillerium** [comment]
Tests are failing right now because they work, investigating the slider tracking again.

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 32]
```diff
@@ -155,15 +168,35 @@ protected override void Update()
...
         }
 
         public void UpdateProgress(double completionProgress)
         {
             Position = slider.CurvePositionAt(completionProgress);
         }
+
+        public bool OnPressed(OsuAction action)
+        {
+            if (cursorTrackingBall && (Hit?.Invoke() ?? false))
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: you're not checking the action type here. while this works at the moment (the only members of `OsuAction` is `LeftButton` and `RightButton`), but this is not a guarantee. That's why the check is in the `Update` condition.

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 33]
```diff
@@ -155,15 +168,35 @@ protected override void Update()
...
+                trackingActions.Add(action);
+                return true;
+            }
+
+            return false;
+        }
+
+        public bool OnReleased(OsuAction action)
+        {
+            if (trackingActions.Contains(action))
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: this is pretty redundant. just call remove.

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 34]
```diff
@@ -155,15 +168,35 @@ protected override void Update()
             {
                 // Make sure to use the base version of ReceivePositionalInputAt so that we correctly check the position.
                 Tracking = canCurrentlyTrack
-                           && lastScreenSpaceMousePosition.HasValue
-                           && ReceivePositionalInputAt(lastScreenSpaceMousePosition.Value)
-                           && (drawableSlider?.OsuActionInputManager?.PressedActions.Any(x => x == OsuAction.LeftButton || x == OsuAction.RightButton) ?? false);
+                        && cursorTrackingBall
+                        && (drawableSlider?.OsuActionInputManager?.PressedActions.Any(x => (x == OsuAction.LeftButton || x == OsuAction.RightButton) && trackingActions.Contains(x)) ?? false);
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: This is still regressing slider behaviour. What about the case where you click outside the slider then drag inside it? Also, there's no test for this.

File: osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs [hunk 35]
```diff
@@ -0,0 +1,233 @@
...
+        {
+            return lastResult.Type == HitResult.Meh;
+        }
+
+        private bool assertGreatJudge()
+        {
+            return lastResult.Type == HitResult.Great;
+        }
+
+        private void performStaticInputTest(Dictionary<List<OsuAction>, double> actionsOnSlider)
```
Comment (peppy) [inline osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs]: why not just make the slider position 0 and use replayframes the whole way down?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 36]
```diff
@@ -37,11 +46,20 @@ public Color4 AccentColour
...
         public SliderBall(Slider slider, DrawableSlider drawableSlider = null)
         {
             this.drawableSlider = drawableSlider;
             this.slider = slider;
+            //We no longer care about whether or not the slider is being tracked by the correct key if the head circle was never hit.
+            if (drawableSlider != null)
+                drawableSlider.HeadCircle.OnNewResult += (o, result) =>
+                {
+                    if (result.Type != HitResult.Miss)
+                        headHit = true;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: i'm not sure what this is trying to achieve but it's definitely not right. hitting the head should have no effect on tracking.

**Comment by nyquillerium** [comment]
Still checking some behavior

**Comment by nyquillerium** [comment]
closing to look at again with a different approach

**Comment by nyquillerium** [comment]
I adjusted the approach, this passes all the tests that we had before so hopefully I haven't forgotten anything this time

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs [hunk 37]
```diff
@@ -37,11 +46,16 @@ public Color4 AccentColour
...
         private readonly DrawableSlider drawableSlider;
+        private bool headJudged;
 
         public SliderBall(Slider slider, DrawableSlider drawableSlider = null)
         {
             this.drawableSlider = drawableSlider;
             this.slider = slider;
+
+            if (drawableSlider != null)
+                drawableSlider.HeadCircle.OnNewResult += (o, result) => headJudged = false;
```
Comment (smoogipoo) [inline osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs]: It's never set to true?
